### ISSUE TITLE
Optimization for GrailSort

### ISSUE THREAD

**Issue by MusicTheorist** [Morwenn/cpp-sort#155]
Hey Morwenn,

Grail Sort is looking great!

I actually found an optimization for it about 2 days ago, and it's verified to be working. The algorithm is now 3-4% faster on average now, and all that it took was a little trick with recursion that I learned from Mr. Astrelin's other algorithm, SqrtSort. It's located in the grail_common_sort method, after the insertion sort for small arrays block.

https://github.com/MusicTheorist/Grail-Sorting-for-Java/blob/master/src/javagrailsort/GrailSort.java

- John

EDIT: Scratch the above, thanks for pointing out the change in space complexity. Perhaps the changes to Insertion Sort will still be nice to implement?

```
    private void grailInsertSort(SortType[] arr, int pos, int len) {
        for(int i = 1; i < len; i++) {
            int insertPos = grailBinSearch(arr, pos, pos + i, pos + i, false);

            if(insertPos < i) {
                SortType item = arr[pos + i];

                int shifts = (pos + i) - insertPos;
                switch(shifts) {
                    case 2:  arr[insertPos + 2] = arr[insertPos + 1];
                    case 1:  arr[insertPos + 1] = arr[insertPos];
                             break;
                    default: System.arraycopy(arr, insertPos, arr, insertPos + 1, shifts);
                }
                arr[insertPos] = item;
            }
        }
    }
```

**Comment by Morwenn** [comment]
Hi!

I'm currently finalizing the version 1.5.0 of the library, which I intend to release tomorrow if possible, once I'm done updating the documentation & examples accordingly. I will have a look at this optimization once I am done with the release.

Thanks for the interest :)

**Comment by MusicTheorist** [comment]
Awesome,

Would you also mind if I could get some help with Vergesort? I'm actually trying to implement it in Java with my sorting visualizing program.

**Comment by Morwenn** [comment]
From the SqrtSort README:

> Idea is the same as in GrailSort ( https://github.com/Mrrl/GrailSort ), but exchange buffer and buffer for block tags are allocated in the heap. Resulting algorithm is 50% faster than `std::sort` (when number of keys in array more than sqrt(N)) or `std::stable_sort`.

If this statement holds (which I always assumed it did), cpp-sort can already provide SqrtSort as follows:

```cpp
using sqrt_sorter = grail_sorter<utility::dynamic_buffer<utility::sqrt>>;
```

I kind of designed the buffered sorters with this in mind, so I didn't feel the need to provide an explicit `sqrt_sorter` in the library.

---

Concerning vergesort: it was a nice project to show off, but I kind of expect its performance to be sub-par for some real distributions when the comparison function is more than trivial: for example when it finds runs whose size are a bit less than *n / log n* it ignores them because they're too small to consider, but it still has to scan them, which makes almost *n / log n* additional comparisons that are subsequently dropped. When the comparison is "trivial" like it is for integers, then the branch predictor can assume that the result of the comparison is `true` and optimize the associated `if`, which means that it will be wrong at most O(log n) time thanks to the jumping mechanism in vergesort. However when comparing other kinds of objects, a branch might happen in the comparison itself, and the real cost of runs a bit smaller than *n / log n* will most likely appear.

Unless you know for sure that the collection you're sorting often has runs bigger than *n / log n*, then vergesort is a pessimization over the algorithm it wraps. Even when you know that you can have runs that big, then you probably want to use timsort which is more adaptive wrt the runs size.

tl;dr: I can surely help implementing vergesort, but the algorithm is most likely not worth the trouble.

**Comment by MusicTheorist** [comment]
I'm honestly still interested! I have efficient sorts and impractical sorts included with the program. The performance doesn't matter, I'm more so just interested in a variety of algorithms.

**Comment by Morwenn** [comment]
I can probably help you if you have a theoretical issue then, but I don't know whether there is an elegant way to design algorithms in Java, so I can hardly help with that :)

**Comment by Morwenn** [comment]
It's strange that SqrtSort implements this optimization while GrailSort doesn't, but the SqrtSort code being slightly more recent than the equivalent GrailSort code I would expect that it's meaningful. A difference could because of the complexity: GrailSort without a buffer at all is supposed to run in O(1) space, but introducing recursion makes the space complexity dependent on the recursion condition.

I will probably have to try to understand how the algorithm works before modifying it, otherwise I might have to change the complexity claims of the library.

**Comment by MusicTheorist** [comment]
...Oh no, good catch. Ah, damn. I didn't want to mess with GrailSort's impressive space complexity.

I'm just going revert the changes. However, I am keeping my latest change of making GrailSort's Insertion Sort a Binary Insertion Sort.

**Comment by MusicTheorist** [comment]
I do have certain ideas on improving GrailSort that I would like to implement in the future.

After each iteration of merging pairs together, the internal buffer wastes a lot of time being shifted back to the start of the array. My thought is that the internal buffer can merge blocks together going in both directions in a stable manner. The only time you would have to sift the internal buffer back to the left side of the array is after all pairs are merged and the buffer sits at the right side of the array.

As you can see, I was really trying to get passed the key gathering process with that recursion method. It's currently one of the largest bottlenecks of GrailSort. My new thoughts are to switch it from an optimized gnome sort with a binary search to the rotations seen in WikiSort.

Once I get these ideas figured out in the future, I will let you know and see what you think!

**Comment by Morwenn** [comment]
To be honest, despite some amount of refactoring, I never made the effort to actually understand how GrailSort and WikiSort actually work. I will need to put in that effort someday if I want to appreciate the quality of the optimizations you're proposing :p

**Comment by MusicTheorist** [comment]
Would you like me to attempt to explain how GrailSort works? I'd be more than happy to!

**Comment by Morwenn** [comment]
I intend to read the paper it is based on first in order to understand the stable in-place merging scheme, then if I have questions about the optimizations (both GrailSort and WikiSort seem to have diverged from the original papers), then I won't hesitate to ask questions :)

**Comment by MusicTheorist** [comment]
I'm definitely going to document GrailSort at some point. I think it is a marvelous algorithm, and it has a lot of potential to be improved on. Have fun learning about it!!

**Comment by MusicTheorist** [comment]
Oh yeah, one last thing (sorry to keep bugging you!): I just uploaded a detailed visualization of GrailSort for myself on YouTube. If you're interested in any visual aids, check it out here: https://www.youtube.com/watch?v=U29NB96Y-9w

**Comment by Morwenn** [comment]
Oh, by the way: if you want other fun sorting algorithms to visualize, then you probably want to look at `split_sort` and `poplar_sort`: both should exhibit interesting structures that could show up in visualization. I implemented both from research papers, so I don't think that you will find many other implementations around. Maybe `drop_merge_sort` would be interesting too, which was originally implemented in Rust.

I believe that it's somewhat easier to understand in-place algorithms when visualizing, so for `poplar_sort` I have a O(1) space implementation in another repository (albeit a bit less efficient): https://github.com/Morwenn/poplar-heap

**Comment by MusicTheorist** [comment]
Okay, then! I think I'll swap out Vergesort for Poplar Heap.

**Comment by MusicTheorist** [comment]
Poplar heap is running great! Looks like a cleaner version of Smoothsort.

**Comment by Morwenn** [comment]
You're not far off: apparently the structure of poplar heap was imagined by Dijkstra when he worked on smoothsort before dropping it and concentrating on Leonardo heaps instead. Poplar heap was later reintroduced in *Smoothsort revisited*, so even though the algorithms don't end up using the same kind of heap they were born from the same ideas :)
